src/api/scenarioData/scenarioData.service.ts

```diff
@@ -0,0 +1,379 @@
+    try {
+      const scenarioDataList = await this.scenarioDataModel.find().exec();
+      return scenarioDataList;
+    } catch (error) {
+      throw new Error(`Error getting all scenario data: ${error.message}`);
+    }
+  }
+
+  async getScenarioData(scenarioId: string): Promise<ScenarioData> {
+    try {
+      const scenarioData = await this.scenarioDataModel.findOne({ _id: scenarioId }).exec();
+      if (!scenarioData) {
+        throw new Error('Scenario data not found');
+      }
+      return scenarioData;
+    } catch (error) {
+      throw new Error(`Error getting scenario data: ${error.message}`);
+    }
+  }
+
+  async createScenarioData(scenarioData: ScenarioData): Promise<ScenarioData> {
+    try {
+      const newScenarioData = new this.scenarioDataModel(scenarioData);
+      const response = await newScenarioData.save();
+      return response;
+    } catch (error) {
+      console.error('Error in createScenarioData:', error);
+      throw new Error(`Failed to add scenario data: ${error.message}`);
+    }
+  }
+
+  async updateScenarioData(scenarioId: string, updatedData: Partial<ScenarioData>): Promise<ScenarioData> {
+    try {
+      const existingScenarioData = await this.scenarioDataModel.findOne({ _id: scenarioId }).exec();
+      if (!existingScenarioData) {
+        throw new Error('Scenario data not found');
+      }
+      Object.assign(existingScenarioData, updatedData);
+      return await existingScenarioData.save();
+    } catch (error) {
+      throw new Error(`Failed to update scenario data: ${error.message}`);
+    }
+  }
+
+  async deleteScenarioData(scenarioId: string): Promise<void> {
+    try {
+      const objectId = new Types.ObjectId(scenarioId);
+      const deletedScenarioData = await this.scenarioDataModel.findOneAndDelete({ _id: objectId });
+      if (!deletedScenarioData) {
+        throw new Error('Scenario data not found');
+      }
+      return;
+    } catch (error) {
+      throw new Error(`Failed to delete scenario data: ${error.message}`);
+    }
+  }
+
+  async getScenariosByRuleId(ruleId: string): Promise<ScenarioData[]> {
+    try {
+      return await this.scenarioDataModel.find({ ruleID: ruleId }).exec();
+    } catch (error) {
+      throw new Error(`Error getting scenarios by rule ID: ${error.message}`);
+    }
+  }
+
+  async getScenariosByFilename(goRulesJSONFilename: string): Promise<ScenarioData[]> {
+    try {
+      return await this.scenarioDataModel.find({ goRulesJSONFilename: goRulesJSONFilename }).exec();
+    } catch (error) {
+      throw new Error(`Error getting scenarios by filename: ${error.message}`);
+    }
+  }
+
+  //handle special characters that may be entered by end users
+  replaceSpecialCharacters(input: string, replacement: string): string {
+    const specialChars = /[\n\r\t\f,]/g;
+    return input.replace(specialChars, (match) => {
+      if (match === ',') {
+        return '-';
+      }
+      return replacement;
+    });
+  }
```

Comment:
I'd recommend putting functions like this in a separate util file. This helps keep this file shorter/cleaner and allows you to re-use those util functions elsewhere in future if you ever need them.